libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -457,53 +462,67 @@ method unsubscribe*(g: GossipSub,
 
 method publish*(g: GossipSub,
                 topic: string,
-                data: seq[byte]) {.async.} =
-  await procCall PubSub(g).publish(topic, data)
+                data: seq[byte]): Future[int] {.async.} =
+  # base returns always 0
+  discard await procCall PubSub(g).publish(topic, data)
   trace "about to publish message on topic", name = topic,
                                              data = data.shortLog
 
-  # TODO: we probably don't need to try multiple times
-  if data.len > 0 and topic.len > 0:
-    var peers = g.mesh.getOrDefault(topic)
-    for _ in 0..<5: # try to get peers up to 5 times
-      if peers.len > 0:
-        break
-
-      if topic in g.topics: # if we're subscribed to the topic attempt to build a mesh
-        await g.rebalanceMesh(topic)
-        peers = g.mesh.getOrDefault(topic)
-      else: # send to fanout peers
-        await g.replenishFanout(topic)
-        if topic in g.fanout:
-          peers = g.fanout.getOrDefault(topic)
-          # set the fanout expiry time
-          g.lastFanoutPubSub[topic] = Moment.fromNow(GossipSubFanoutTTL)
+  var peers: HashSet[string]
 
-      # wait a second between tries
-      await sleepAsync(1.seconds)
+  if topic.len > 0: # data could be 0/empty
+    if topic in g.topics: # if we're subscribed use the mesh
+      peers = g.mesh.getOrDefault(topic)
+    else: # not subscribed, send to fanout peers
+      # try optimistically
+      peers = g.fanout.getOrDefault(topic)
+      if peers.len == 0:
+        # ok we had nothing.. let's try replenish inline
+        g.replenishFanout(topic)
+        peers = g.fanout.getOrDefault(topic)
 
     let
       msg = Message.init(g.peerInfo, data, topic, g.sign)
       msgId = g.msgIdProvider(msg)
 
     trace "created new message", msg
+
+    trace "publishing on topic", name = topic, peers = peers
+    if msgId notin g.mcache:
+      g.mcache.put(msgId, msg)
+
     var sent: seq[Future[void]]
     for p in peers:
+      # avoid sending to self
       if p == g.peerInfo.id:
         continue
 
-      trace "publishing on topic", name = topic
-      if msgId notin g.mcache:
-        g.mcache.put(msgId, msg)
-
-      if p in g.peers:
-        sent.add(g.peers[p].send(@[RPCMsg(messages: @[msg])]))
-    checkFutures(await allFinished(sent))
+      let peer = g.peers.getOrDefault(p)
+      if not isNil(peer) and not isNil(peer.peerInfo):
+        trace "publish: sending message to peer", peer = p
+        sent.add(peer.send(@[RPCMsg(messages: @[msg])]))
+      else:
+        # Notice this needs a better fix! for now it's a hack
+        error "publish: peer or peerInfo was nil"
+        if topic in g.mesh:
```

Comment:
we should call `handleDisconnect` here instead, otherwise it will not clear all the tables.

Reply:
Ok sure, to be honest this was a try, I did not trigger it at all yet while running as validator.. not sure why @arnetheduck hit this issue....

Reply:
Ah btw cannot call `handledisconnect` cos peer does not exist its nil in `g.peers`

Reply:
anyway there is a bug here and it messes up things.. this code will go away once fixes